### ISSUE TITLE
Content Delivery API - Missing Accept-Language in Vary header

### ISSUE THREAD

**Issue by jdpnielsen** [umbraco/Umbraco-CMS#17034]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

13.4.1

### Bug summary

I'm using the Content Delivery API with a multi-language site. When getting content in a specific language i use the `Accept-Language` header as specified in the [documentation](https://docs.umbraco.com/umbraco-cms/reference/content-delivery-api#concepts). Unfortunately, this can lead to issues with browser caching of content, due to the response being cached and served across any language variation. This happens because the browser does not take request headers into account when determining matching cache entries.

### Specifics

Screen recording shows issue with browser caching of requests, using the appended reproduction example:

https://github.com/user-attachments/assets/ef4d828b-432f-47bb-8023-ad4b516459ac

```html
<!DOCTYPE html>
<html>
<head>
  <title>Home</title>
  <script type="module">
    document.getElementById('load-english-content').addEventListener('click', async () => {
      const data = await fetchByLanguage('en-US');

      updateResultDiv('english-content', data);
    });

    document.getElementById('load-danish-content').addEventListener('click', async () => {
      const data = await fetchByLanguage('da-DK');

      updateResultDiv('danish-content', data);
    });

    async function fetchByLanguage(language) {
      const myHeaders = new Headers();
      myHeaders.append("Accept", "application/json");
      myHeaders.append("Accept-Language", language);

      try {
        const response = await fetch('https://localhost:44321/umbraco/delivery/api/v2/content?filter=contentType:eventEntity', {
          method: "GET",
          cache: 'force-cache',
          headers: myHeaders,
          redirect: "follow"
        });

        const result = await response.json();
        return result;
      } catch (error) {
        console.error(error);
        throw error;
      };
    }

    function updateResultDiv(divId, data) {
      const div = document.getElementById(divId);
      div.innerHTML = '';
      const table = div.appendChild(document.createElement('table'));

      table.appendChild(document.createElement('th')).textContent = 'Name';
      table.appendChild(document.createElement('th')).textContent = 'URL';

      data.items.forEach(item => {
        const row = document.createElement('tr');

        const cell1 = row.appendChild(document.createElement('td'));
        cell1.textContent = item.name;
        const cell2 = row.appendChild(document.createElement('td'));
        cell2.textContent = item.url;

        table.appendChild(row);
      });
    }
  </script>
</head>
<body>
  <div>
    <h1>English content</h1>
    <button id="load-english-content">Load content</button>
    <div id="english-content"></div>
  </div>
  <div>
    <h1>Danish content</h1>
    <button id="load-danish-content">Load content</button>
    <div id="danish-content"></div>
  </div>
</body>
</html>
```

### Steps to reproduce

1. Setup content with language variants
2. Whitelist frontend Cors origin
3. Using the reproduction example above:
     1. Request one language using `Accept-Language` header to get content in a specific language.
     2. In the same browser session, repeat the request with a different `Accept-Language` header
4. Browser cache will return content matching the first request, instead of the second

### Expected result / actual result

A standard solution to this is for the server to respond to a request with the [Vary header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary). Unfortunately it seems to be configured to respond with `Vary: Accept-Encoding` (At least on Umbraco Cloud). 

I propose two solutions: 

1. Include `Accept-Language`, `Preview` and `Start-Item` in the `Vary` header where applicable.
2. Make all applicable inputs to the content query part of the url. Ie. `https://localhost:44321/umbraco/delivery/api/v2/content?filter=contentType:eventEntity&language=da-DK`

**Comment by jdpnielsen** [comment]
Not directly related to this issue is https://github.com/umbraco/Umbraco.Cloud.Issues/discussions/815, which also requests improvements to `Vary` header handling.

**Comment by gogolineq** [comment]
@jdpnielsen did you find solution for this issue? I have the same problem - Accept-Language is not working and content is cached via all cultures.

**Comment by jdpnielsen** [comment]
> @jdpnielsen did you find solution for this issue? I have the same problem - Accept-Language is not working and content is cached via all cultures.

@gogolineq Yes, actually. If you add a fake query parameter to your request, the browser will correctly match the cache entry. Ie. something like this:

`/umbraco/delivery/api/v2/content?filter=contentType:eventEntity&language=da-DK`

**Comment by gogolineq** [comment]
@jdpnielsen it's working, thanks!

**Comment by kjac** [comment]
Hi everyone,

Thank you for this ... and sorry for the belated response. 

I think this would be a great addition. But as this is a feature request, I will move this to the discussions for the time being 👍 

You can roll your own solution with a piece of middleware. This way you don't have to add additional query parameters to the API calls:

```csharp
public class AddVaryHeaderMiddleware
{
    private readonly RequestDelegate _next;

    public AddVaryHeaderMiddleware(RequestDelegate next)
        => _next = next;

    public async Task InvokeAsync(HttpContext context)
    {
        // add the Vary headers you need
        context.Response.Headers["Vary"] = "Accept, Accept-Language";
        await _next(context);
    }
}
```
